org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/ILanguageServerAccess.java

```diff
@@ -0,0 +1,158 @@
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.xtext.ide.server;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+import org.eclipse.emf.common.util.URI;
+import org.eclipse.emf.ecore.resource.Resource;
+import org.eclipse.emf.ecore.resource.ResourceSet;
+import org.eclipse.lsp4j.InitializeParams;
+import org.eclipse.lsp4j.InitializeResult;
+import org.eclipse.lsp4j.services.LanguageClient;
+import org.eclipse.xtext.ide.serializer.IChangeSerializer;
+import org.eclipse.xtext.resource.IResourceDescription;
+import org.eclipse.xtext.resource.IResourceDescriptions;
+import org.eclipse.xtext.resource.impl.ResourceDescriptionsProvider;
+import org.eclipse.xtext.util.CancelIndicator;
+import org.eclipse.xtext.xbase.lib.util.ToStringBuilder;
+
+/**
+ * API for interacting with a running language server.
+ * 
+ * @author Sven Efftinge - Initial contribution and API
+ */
+public interface ILanguageServerAccess {
+	class Context {
+		private final Resource resource;
+
+		private final Document document;
+
+		private final boolean isDocumentOpen;
+
+		private final CancelIndicator cancelChecker;
+
+		public Context(Resource resource, Document document, boolean isDocumentOpen, CancelIndicator cancelChecker) {
+			this.resource = resource;
+			this.document = document;
+			this.isDocumentOpen = isDocumentOpen;
+			this.cancelChecker = cancelChecker;
+		}
+
+		@Override
+		public String toString() {
+			ToStringBuilder b = new ToStringBuilder(this);
+			b.add("resource", resource);
+			b.add("document", document);
+			b.add("isDocumentOpen", isDocumentOpen);
+			b.add("cancelChecker", cancelChecker);
+			return b.toString();
+		}
+
+		public Resource getResource() {
+			return resource;
+		}
+
+		public Document getDocument() {
+			return document;
+		}
+
+		public boolean isDocumentOpen() {
+			return isDocumentOpen;
+		}
+
+		public CancelIndicator getCancelChecker() {
+			return cancelChecker;
+		}
+	}
+
+	class IndexContext {
+		private final IResourceDescriptions index;
+
+		private final CancelIndicator cancelChecker;
+
+		public IndexContext(IResourceDescriptions index, CancelIndicator cancelChecker) {
+			this.index = index;
+			this.cancelChecker = cancelChecker;
+		}
+
+		@Override
+		public String toString() {
+			ToStringBuilder b = new ToStringBuilder(this);
+			b.add("index", index);
+			b.add("cancelChecker", cancelChecker);
+			return b.toString();
+		}
+
+		public IResourceDescriptions getIndex() {
+			return index;
+		}
+
+		public CancelIndicator getCancelChecker() {
+			return cancelChecker;
+		}
+	}
+
+	interface IBuildListener {
+		void afterBuild(List<IResourceDescription.Delta> deltas);
+	}
+
+	/**
+	 * provides read access to a fully resolved resource and Document.
+	 */
+	<T extends Object> CompletableFuture<T> doRead(String uri, Function<ILanguageServerAccess.Context, T> function);
```

Comment:
Stylistic: `T extends Object` is redundant.

Also: this should probably be `Function<? super ILSA.Context, ? extends T>`

Reply:
will have a look

Reply:
this would break clients like LanguageServerImpl